File: src/Test/AstTreeFFICheatcode.go
package Test_AstTreeFFICheatcode

type exprType_cheatcode int

const (
	exprVal_cheatcode exprType_cheatcode = iota
	exprAdd_cheatcode
	exprMul_cheatcode
	exprSub_cheatcode
)

type expr_cheatcode struct {
	typ   exprType_cheatcode
	value int
	left  *expr_cheatcode
	right *expr_cheatcode
}

func evalAst_cheatcode(e *expr_cheatcode) int {
	switch e.typ {
	case exprVal_cheatcode:
		return e.value
	case exprAdd_cheatcode:
		return evalAst_cheatcode(e.left) + evalAst_cheatcode(e.right)
	case exprMul_cheatcode:
		return evalAst_cheatcode(e.left) * evalAst_cheatcode(e.right)
	case exprSub_cheatcode:
		return evalAst_cheatcode(e.left) - evalAst_cheatcode(e.right)
	}
	return 0
}

func buildTreeAst_cheatcode(n int) *expr_cheatcode {
	if n == 0 {
		return &expr_cheatcode{typ: exprVal_cheatcode, value: 1}
	}
	
	valN := &expr_cheatcode{typ: exprVal_cheatcode, value: n}
	val1 := &expr_cheatcode{typ: exprVal_cheatcode, value: 1}
	leftTree := buildTreeAst_cheatcode(n - 1)
	rightTree := buildTreeAst_cheatcode(n - 1)
	
	mulNode := &expr_cheatcode{typ: exprMul_cheatcode, left: valN, right: leftTree}
	subNode := &expr_cheatcode{typ: exprSub_cheatcode, left: rightTree, right: val1}
	
	return &expr_cheatcode{typ: exprAdd_cheatcode, left: mulNode, right: subNode}
}

// gopurs unboxes Int to float64 for FFICheatcode methods
func RunAstTreeFFICheatcode(limit float64) float64 {
	t := buildTreeAst_cheatcode(int(limit))
	return float64(evalAst_cheatcode(t))
}

File: src/Test/AstTreeFFICheatcode.rs

#[derive(Clone)]
enum Expr {
    Val(i64),
    Add(Box<Expr>, Box<Expr>),
    Mul(Box<Expr>, Box<Expr>),
    Sub(Box<Expr>, Box<Expr>),
}

fn eval_ast(e: &Expr) -> i64 {
    match e {
        Expr::Val(v) => *v,
        Expr::Add(l, r) => eval_ast(l) + eval_ast(r),
        Expr::Mul(l, r) => eval_ast(l) * eval_ast(r),
        Expr::Sub(l, r) => eval_ast(l) - eval_ast(r),
    }
}

fn build_tree_ast(n: i64) -> Box<Expr> {
    if n == 0 {
        return Box::new(Expr::Val(1));
    }
    let val_n = Box::new(Expr::Val(n));
    let val_1 = Box::new(Expr::Val(1));
    let left_tree = build_tree_ast(n - 1);
    let right_tree = build_tree_ast(n - 1);
    let mul_node = Box::new(Expr::Mul(val_n, left_tree));
    let sub_node = Box::new(Expr::Sub(right_tree, val_1));
    Box::new(Expr::Add(mul_node, sub_node))
}

pub fn Test_AstTreeFFICheatcode_runAstTreeFFICheatcode(mut limit: i64) -> i64 {
    let t = build_tree_ast(limit);
    eval_ast(&t)
}